File: jpp/examples/websocket_client_example.py
#!/usr/bin/env python3
"""
Example Python client for listening to JPP invoice payment updates via WebSocket.

Usage:
    python websocket_client_example.py <invoice_uuid>

Example:
    python websocket_client_example.py 1234567890abcdef1234567890abcdef

Requirements:
    pip install websocket-client
"""

import sys
import json
import websocket
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def on_message(ws, message):
    """Handle incoming WebSocket messages."""
    data = json.loads(message)
    logger.info("=" * 60)
    logger.info("PAYMENT UPDATE RECEIVED!")
    logger.info("=" * 60)
    logger.info(f"Type: {data.get('type')}")
    logger.info(f"Transaction ID: {data.get('txid')}")
    logger.info(f"Paid At: {data.get('paid_at')}")
    if data.get('memo'):
        logger.info(f"Memo: {data.get('memo')}")
    
    invoice = data.get('invoice', {})
    if invoice:
        logger.info(f"\nInvoice Details:")
        logger.info(f"  Payment ID: {invoice.get('payment_id')}")
        logger.info(f"  Network: {invoice.get('network')}")
        logger.info(f"  Currency: {invoice.get('currency')}")
        logger.info(f"  Total BCH: {invoice.get('payment', {}).get('total_bch', 'N/A')}")
    
    logger.info("=" * 60)


def on_error(ws, error):
    """Handle WebSocket errors."""
    logger.error(f"WebSocket error: {error}")


def on_close(ws, close_status_code, close_msg):
    """Handle WebSocket connection close."""
    logger.info(f"WebSocket connection closed (status: {close_status_code}, msg: {close_msg})")


def on_open(ws):
    """Handle WebSocket connection open."""
    logger.info("WebSocket connection established")
    logger.info("Listening for payment updates...")


def main():
    if len(sys.argv) < 2:
        print("Usage: python websocket_client_example.py <invoice_uuid>")
        print("Example: python websocket_client_example.py 1234567890abcdef1234567890abcdef")
        sys.exit(1)
    
    invoice_uuid = sys.argv[1]
    
    # Default to local development, change to your server URL
    host = sys.argv[2] if len(sys.argv) > 2 else "localhost:8000"
    protocol = "wss" if "watchtower.cash" in host else "ws"
    
    ws_url = f"{protocol}://{host}/ws/jpp/invoice/{invoice_uuid}/"
    
    logger.info(f"Connecting to: {ws_url}")
    
    ws = websocket.WebSocketApp(
        ws_url,
        on_open=on_open,
        on_message=on_message,
        on_error=on_error,
        on_close=on_close
    )
    
    # Run forever
    ws.run_forever()


if __name__ == "__main__":
    main()

File: jpp/examples/websocket_client_example.js
#!/usr/bin/env node
/**
 * Example Node.js client for listening to JPP invoice payment updates via WebSocket.
 *
 * Usage:
 *     node websocket_client_example.js <invoice_uuid> [host]
 *
 * Example:
 *     node websocket_client_example.js 1234567890abcdef1234567890abcdef
 *     node websocket_client_example.js 1234567890abcdef1234567890abcdef watchtower.cash
 *
 * Requirements:
 *     npm install ws
 */

const WebSocket = require('ws');

function main() {
    const args = process.argv.slice(2);
    
    if (args.length < 1) {
        console.log('Usage: node websocket_client_example.js <invoice_uuid> [host]');
        console.log('Example: node websocket_client_example.js 1234567890abcdef1234567890abcdef');
        console.log('Example: node websocket_client_example.js 1234567890abcdef1234567890abcdef watchtower.cash');
        process.exit(1);
    }
    
    const invoiceUuid = args[0];
    const host = args[1] || 'localhost:8000';
    const protocol = host.includes('watchtower.cash') ? 'wss' : 'ws';
    
    const wsUrl = `${protocol}://${host}/ws/jpp/invoice/${invoiceUuid}/`;
    
    console.log(`Connecting to: ${wsUrl}`);
    
    const ws = new WebSocket(wsUrl);
    
    ws.on('open', function() {
        console.log('WebSocket connection established');
        console.log('Listening for payment updates...');
    });
    
    ws.on('message', function(data) {
        const message = JSON.parse(data.toString());
        
        console.log('='.repeat(60));
        console.log('PAYMENT UPDATE RECEIVED!');
        console.log('='.repeat(60));
        console.log(`Type: ${message.type}`);
        console.log(`Transaction ID: ${message.txid}`);
        console.log(`Paid At: ${message.paid_at}`);
        
        if (message.memo) {
            console.log(`Memo: ${message.memo}`);
        }
        
        if (message.invoice) {
            console.log('\nInvoice Details:');
            console.log(`  Payment ID: ${message.invoice.payment_id}`);
            console.log(`  Network: ${message.invoice.network}`);
            console.log(`  Currency: ${message.invoice.currency}`);
            
            if (message.invoice.payment) {
                console.log(`  Total BCH: ${message.invoice.payment.total_bch || 'N/A'}`);
            }
        }
        
        console.log('='.repeat(60));
    });
    
    ws.on('error', function(error) {
        console.error(`WebSocket error: ${error.message}`);
    });
    
    ws.on('close', function(code, reason) {
        console.log(`WebSocket connection closed (code: ${code}, reason: ${reason.toString()})`);
    });
}

main();

